### ISSUE TITLE
AIState does not update correctly with parallel StreamUI calls

### ISSUE THREAD

**Issue by ZacharyHangoc** [vercel/ai#3084]
### Description

I am getting this problem when calling a server actions streamUI function using useActions hook and AI context.  I am sending an array of instructions and calling multiple streamUI in parallel  (Promise.all) to generate a few components to be returned to the client. The problem is the state does not update correctly. I have to refresh the page to get the correct state (The saving to the database using onSetAIState works fine)

Steps:

1. setup useActions hook in client component and call it from async click handler, sending an array of instructions
2. streamExercises server action which takes the array and awaits Promise.all with the loop of the instructions
3. in each LLM tool, call aiState.done with updated state
4. At the end return responses and aiState.get()
5. aiState.get() returned is empty

### Code example

`export async function streamExercises({ path }: { path: any }) {
  "use server";
  const aiState = getMutableAIState<typeof AI>();

  const { pathContent } = aiState.get();
const exerciseInstructions = [
  {
    id: 1,
    task: "Generate a short explanation of the topic.",
  },
  {
    id: 2,
    task: "Create a multiple-choice question about the topic.",
  },
  {
    id: 3,
    task: "Fetch a visual representation of the topic.",
  },
];

const responses = await Promise.all(
    pathContent.flatMap((section, sectionIndex) =>
      exerciseInstructions.map(async (instruction, instructionIndex) => {
        const response = await streamUI({
          initial: <Loader2 className="animate-spin w-[30px] h-[30px]" />,
          toolChoice: "auto",
          model: model,
          maxTokens: 500,
          maxRetries: 0,
          system: `AI teaching assistant. Generate relevant, interactive learning content using available tools and using Source Documents if provided:
- createMultipleChoiceQuestion for quizzes
- createShortExplanation for concise, markdown-formatted learning content
- createVisual for visual aids
Adapt content to specified grade level and difficulty. If task is impossible, state limitations.`,
          prompt: `Task: ${instruction.task}
Path: ${title} | ${description} | Grade: ${grade}
Module: ${section.moduleTitle} | ${section.moduleDescription}
Section: ${JSON.stringify(section)}
Generate content tailored to grade level: ${grade} and difficulty: ${level}`,tools: {
            createShortExplanation: {
              description:
                "Create a short topic learning section optimized to help a student understand and learn. Use this to generate a short learning section adapted to the student's grade level. MUST be in markdown format. Header is always H2 (##).",

              parameters: z.object({
                content: z.string(),
                // .describe(
                //   "A simple to understand, concise learning section. Adapted to the student's grade level."
                // ),
              }),
              generate: async function* ({ content }) {
                // yield <LoadingCard />;
                const componentId = generateId();
                const toolCallId = generateId();

                aiState.done({
                  ...aiState.get(),
                  pathContent: aiState
                    .get()
                    .pathContent.map((pathSection: any) => {
                      if (
                        pathSection.id === section.id &&
                        pathSection.module_id === section.module_id
                      ) {
                        return {
                          ...pathSection,
                          content: [
                            ...(pathSection.content || []),
                            {
                              id: componentId,
                              role: "tool",
                              order: instructionIndex,
                              type: "tool-result",
                              toolName: "createShortExplanation",
                              toolCallId,
                              result: { id: componentId, content },
                            },
                          ],
                        };
                      }
                      return pathSection;
                    }),
                });

                return (
                  <AssistantMessageEdit
                    props={{ id: componentId, editorMode: true, content }}
                  />
                );
              },
            },
            otherTools....
            }
  return {
          id: `${Date.now()}-${sectionIndex}-${instructionIndex}`,
          role: "assistant",
          sectionId: section.id,
          moduleId: section.module_id,
          instruction: instruction.task,
          content: response.value,
        };
 })
    )
  );

  // Get the final state after all updates

  // Get the final state after all updates
  const finalState = aiState.get();
  console.log(finalState);

  return { content: responses.flat(), aiState: finalState };
}`

### Additional context

On the client:

`const handleStreamContent = async () => {
    onOpen();

    try {
      const { content, aiState: newAIState } = await streamExercises({
        path: pathData,
      });

      console.log(content);

      setUIState((prev: any) => {
        const updatedContent = prev.map((section: any) => {
          const matchingContent = content.filter(
            (item: any) =>
              item.sectionId === section.id &&
              item.moduleId === section.module_id
          );

          if (matchingContent.length > 0) {
            return {
              ...section,
              content: [...section.content, ...matchingContent],
            };
          }
          return section;
        });

        return updatedContent;
      });
      console.log(newAIState.pathContent);

      setAIState(() => newAIState);
    } catch (error) {
      console.log(error);
      console.error("Error generating content:", error);
    } finally {
      onClose();
    }
  };`

**Comment by aayush-kapoor** [comment]
Marking this as closed for now - if you face any regression or have a similar problem after migrating to the latest AI SDK version, please feel free to open a new issue (with possible reproduction)

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Bug**  
Confidence: **Medium**

### Rationale
The issue reports a concrete failure in existing AI state update behavior during parallel `streamUI` operations, although it lacks a confirmed reproduction on the latest SDK.

### Evidence
- Parallel `streamUI` calls invoke `aiState.done(...)`, but `aiState.get()` remains empty after `Promise.all` completes.
- The persisted database state is correct and refreshing restores the expected state, indicating inconsistent in-memory/client AI state behavior.
- The closing comment does not identify misuse or a feature request; it asks for a new report if the problem persists on the latest SDK.
